tests/unit_tests/ut_midi_stream.cpp

```diff
@@ -0,0 +1,84 @@
+#define CATCH_CONFIG_MAIN
+
+#include <catch2/catch.hpp>
+#include <composer/midi_stream.h>
+#include <MidiFile.h>
+
+using cycfi::q::literals::operator""_ms;
+using cycfi::q::literals::operator""_s;
+using cycfi::q::literals::operator""_dB;
+
+using namespace composer;
+
+TEST_CASE ("MidiEventList contains harmony") {
+
+    // Empty MIDI file
+    smf::MidiFile midi_file;
+
+    REQUIRE_THROWS_WITH(is_harmony(midi_file[0]), "Empty MidiEventList.");
+
+    // MIDI file containing melody
+    midi_file.read("../resources/twinkle.midi");
+
+    if (midi_file.getTrackCount() > 1) {
+        midi_file.joinTracks();
+    }
+
+    REQUIRE(is_harmony(midi_file[0]) == false);
+
+    // MIDI file containing harmony
+    midi_file.read("../resources/twinkle_harmonic.mid");
+
+    if (midi_file.getTrackCount() > 1) {
+        midi_file.joinTracks();
+    }
+
+    REQUIRE(is_harmony(midi_file[0]));
+}
+
+TEST_CASE ("Dispatch MIDI messages to synthesizer") {
+
+    // Test resources
+    auto env_cfg = cycfi::q::envelope::config {
+        100_ms    // attack rate
+        , 1_s     // decay rate
+        , -12_dB // sustain level
+        , 5_s    // sustain rate
+        , 1_s    // release rate
+    };
+
+    synth synth_(env_cfg);
+    midi_processor processor(synth_);
+    smf::MidiFile midi_file;
+
+    REQUIRE_THROWS_WITH(dispatch_midi_messages(processor, midi_file[0]), "Empty MidiEventList.");
+
+    // MIDI file containing melody
+    midi_file.read("../resources/twinkle.midi");
+
+    if (midi_file.getTrackCount() > 1) {
+        midi_file.joinTracks();
+    }
+
+    REQUIRE_NOTHROW(dispatch_midi_messages(processor, midi_file[0]));
```

Comment:
Mesmo. De modo geral, casos extremos precisam ser tratados de maneira que corresponda a API da função em vez de lançarem exceção.

Reply:
Eu fiquei em dúvida nesse teste, porque `dispatch_midi_messages` não tem retorno. Como está testando o caso de passar uma melodia para a função, eu posso manter o REQUIRE_NOTHROW para esperar que a função não vai lançar a exceção do is_harmony?

Reply:
Pode. Mas o is_harmony não deveria estar lançando exceção também. :P

Reply:
Se já se sabe que a função não vai lançar exceção nunca, você pode colocar `noexcept` nela que o compilador já te proíbe de colocar exceção nela. Não precisa nem testar.

Reply:
Se bem que estou vendo que tem o caso onde não é harmonia e vai lançar exceção por questão de implementação, então deixa assim mesmo.